### PR TITLE
Feature/ego core

### PR DESCRIPTION
**Pull request by ra-ver**
Ego can now:
1. Add/List/Remove users, groups and applications
2. Add/Remove user to groups
3. Add/Remove applications to groups
4. Add/Remove applications access to users

### CODE REVIEW

File: src/main/java/org/overture/ego/model/QueryInfo.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+   /*
+      Un-sanitized get will return empty string
+    */
+   public String getSort(){
+      return this.getSort(s -> "");
+   }
+
+   /*
+    Using string templates with JDBI opens up the room for SQL Injection
```
Comment (andricDu) [inline src/main/java/org/overture/ego/model/QueryInfo.java]: what is the necessity for using string templates and `@Define` over a `@Bind`

**Comment by ra-ver** [comment]
Merged refactoring changes in order to fix the comments related to data access layer

File: src/main/java/org/overture/ego/security/JWTAccessToken.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+  @Override
+  public Map<String, Object> getAdditionalInformation() {
+    val output = new HashMap<String, Object>();
+    output.put("groups",getUser().get("groups"));
+    return output;
+  }
+
+  @Override
+  public Set<String> getScope() {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/security/JWTAccessToken.java]:
```java
import static com.google.common.collect.Sets.newHashSet;



  @Override
  public Set<String> getScope() {
    return newHashSet((ArrayList<String>)getUser().get("roles"));
  }
```

Though to be honest, I'm seeing a lot of `Map<String, Object>` in the code. Does it make sense to create `Value` classes to handle this so we are least get type safety? 

A new developer coming in seeing `Map<String,Object>` being thrown around in the code won't know what to expect of use unless they drop down into a debugger.

File: src/main/java/org/overture/ego/security/JWTAccessToken.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    return (int) ((System.currentTimeMillis() - tokenClaims.getExpiration().getTime()) / 1000L);
+  }
+
+  @Override
+  public String getValue() {
+    return token;
+  }
+
+  private Map getUser(){
+    return (Map)((Map)tokenClaims.get("context")).get("user");
```
Comment (andricDu) [inline src/main/java/org/overture/ego/security/JWTAccessToken.java]: oh boy, no type knowledge about the map? 

The ghost of type safety is displeased. 👻

File: src/main/java/org/overture/ego/service/GroupService.java [hunk 4]
```diff
@@ -16,8 +16,121 @@
...
+    });
+    groupRepository.save(group);
+  }
+
+  public Group get(String groupId) {
+    //TODO: change id to string
+    return groupRepository.findOne(Integer.parseInt(groupId));
+  }
+
+  public Group getByName(String groupName) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/GroupService.java]: NonNull

File: src/main/java/org/overture/ego/service/GroupService.java [hunk 5]
```diff
@@ -16,8 +16,121 @@
...
+    return groupRepository.findOneByNameIgnoreCase(groupName);
+  }
+
+  public Group update(Group updatedGroupInfo) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/GroupService.java]: NonNull

File: src/main/java/org/overture/ego/service/GroupService.java [hunk 6]
```diff
@@ -16,8 +16,121 @@
...
+    return groupRepository.save(updatedGroupInfo);
+  }
+
+  public void delete(String groupId) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/GroupService.java]: NonNull

File: src/main/java/org/overture/ego/service/GroupService.java [hunk 7]
```diff
@@ -16,8 +16,121 @@
...
+    //TODO: change id to string
+     groupRepository.delete(Integer.parseInt(groupId));
+  }
+
+  public Page<Group> listGroups(Pageable pageable) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/GroupService.java]: NonNull

File: src/main/java/org/overture/ego/service/GroupService.java [hunk 8]
```diff
@@ -16,8 +16,121 @@
...
+    return groupRepository.findAll(pageable);
+  }
+
+  public Page<Group> findGroups(String query, Pageable pageable) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/GroupService.java]: NonNull

File: src/main/java/org/overture/ego/service/GroupService.java [hunk 9]
```diff
@@ -16,8 +16,121 @@
...
+    return groupRepository.findAll(GroupSpecification.containsText(query), pageable);
+  }
+
+  public Page<Group> findUsersGroup(String userId, Pageable pageable){
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/GroupService.java]: NonNull

File: src/main/java/org/overture/ego/service/UserService.java [hunk 10]
```diff
@@ -16,45 +16,122 @@
...
   public User get(String userId) {
-    if (userRepository.read(userId) == null || userRepository.read(userId).size() == 0)
-      return null;
-    else
-      return userRepository.read(userId).get(0);
+    //TODO: change id to string
+    return userRepository.findOne(Integer.parseInt(userId));
+  }
+
+  public User getByName(String userName) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/UserService.java]: NonNull

File: src/main/java/org/overture/ego/service/UserService.java [hunk 11]
```diff
@@ -16,45 +16,122 @@
...
-    return updatedUserInfo;
+    return userRepository.save(updatedUserInfo);
   }
 
   public void delete(String userId) {
-    userRepository.delete(userId);
+    userRepository.delete(Integer.parseInt(userId));
+  }
+
+  public Page<User> listUsers(Pageable pageable) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/UserService.java]: NonNull

File: src/main/java/org/overture/ego/service/UserService.java [hunk 12]
```diff
@@ -16,45 +16,122 @@
...
+    return userRepository.findAll(pageable);
+  }
+
+  public Page<User> findUsers(String query, Pageable pageable) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/UserService.java]: NonNull

File: src/main/java/org/overture/ego/service/UserService.java [hunk 13]
```diff
@@ -16,45 +16,122 @@
...
+    return userRepository.findAll(UserSpecification.containsText(query), pageable);
+  }
+
+  public void deleteUserFromGroup(String userId, List<String> groupIDs) {
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/UserService.java]: NonNull

File: src/main/java/org/overture/ego/service/UserService.java [hunk 14]
```diff
@@ -16,45 +16,122 @@
...
+    userRepository.save(user);
+  }
+
+  public Page<User> findGroupsUsers(String groupId, Pageable pageable){
+    return userRepository.findAll(
+            UserSpecification.inGroup(Integer.parseInt(groupId)),
+            pageable);
+  }
+
+  public Page<User> findGroupsUsers(String groupId, String query, Pageable pageable){
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/UserService.java]: NonNull

File: src/main/java/org/overture/ego/service/UserService.java [hunk 15]
```diff
@@ -16,45 +16,122 @@
...
+  public void deleteUserFromApp(String userId, List<String> appIDs) {
+    //TODO: change id to string
+    val user = userRepository.findOne(Integer.parseInt(userId));
+    appIDs.forEach(appId -> {
+      user.removeApplication(Integer.parseInt(appId));
+    });
+    userRepository.save(user);
+  }
+
+  public Page<User> findGroupsUsers(String groupId, Pageable pageable){
```
Comment (andricDu) [inline src/main/java/org/overture/ego/service/UserService.java]: NonNull

File: src/main/resources/application.yml [hunk 16]
```diff
@@ -25,4 +18,51 @@ spring.datasource:
...
   max-active: 10
   max-idle: 1
-  min-idle: 1
\ No newline at end of file
+  min-idle: 1
+
+# Facebook Connection Details
+facebook:
+  client:
+    id: 140524976574963
```
Comment (andricDu) [inline src/main/resources/application.yml]: you okay with this stuff public?

**Comment by andricDu** [comment]
Another comment is having a catch all `Utils.java` class is usually bad practice as it encourages no cohesion in the code base. 

What is preferable is a `Util` package where you have utility classes of high cohesion of the static helper methods inside. 

Like:
```
Util
|-Files.java
|-Responses.java
|-Exceptions.java
|-Parsers.java
|-Collections.java
```

Just as an example.

File: src/main/resources/application.yml (hunk 16 above)
Comment (joneubank) [reply src/main/resources/application.yml]: Similarly, the google client-ids probably shouldn't be in a public repo...

Comment (cheapsteak) [reply src/main/resources/application.yml]: @Jephuff could you please regenerate the facebook app credentials?

**Pull request merged**